File: Figure 2/run_spinodal_decomp_HPC.jl
using DelimitedFiles
using Dates
using BenchmarkTools
using Printf
# import Pkg; Pkg.add("FFTW")
# Pkg.add("Plots")
# Pkg.add("CSV")
# Pkg.add("LinearAlgebra")
# Pkg.add("BenchmarkTools")
# Pkg.add("StaticArrays")
# Pkg.add("Printf")

include("CahnHilliard_NMG.jl")
include("CahnHilliard_SAV.jl")
include("ch_movie_from_file.jl")

indir = "./IC/"
outdir = "/project/g_bme-janeslab/SarahG/spinodal_decomp_06_2025"

GridSize = parse.(Int, ARGS[1])
boundary = ARGS[2]
print = lowercase(ARGS[3]) == "true"
method = ARGS[4]
note= ARGS[5]
SLURM_ID=ARGS[6]
# if note == "50p"
#     note = ""
# end

nx = GridSize
ny = nx

n_relax = 4
h = 1 / GridSize
m = 8
epsilon = m * h / (2 * sqrt(2) * atanh(0.9))

printphi = true
#use neumann-smoothed IC for both periodic and neumann sims
phi0 = initialization_from_file("$(indir)initial_phi_$(GridSize)_smooth_n_relax_$(n_relax)_$(note)_neumann.csv", GridSize, GridSize, delim=',', transpose_matrix=false)
dt = 5.5e-7
max_it = 20000

if print
    dt_out = 10
else
    dt_out = 20000
end

id = @sprintf("%s_Julia_%d_dt_%.2e_Nx_%d_%s_dtout_%d%s", method, max_it, dt, GridSize, boundary, dt_out, note)
pathname = @sprintf("%s/out_julia/%s_", outdir, id)
date_time = now()

if method == "NMG"
    result, elapsed_time, mem_allocated, gc_time, memory_counters = @timed CahnHilliard_NMG(phi0; t_iter=max_it, dt=dt, dt_out=dt_out, m=m, boundary=boundary, printphi=printphi, pathname=pathname)
    open("$(outdir)/Job_specs_updated.csv", "a", lock=false) do f
        writedlm(f, [Dates.format(date_time, "mm/dd/yyyy HH:MM:SS") "Julia" method GridSize epsilon dt 1e-5 max_it 1e4 dt_out print boundary pathname elapsed_time (mem_allocated/1e6) SLURM_ID note], ",")
    end
elseif method == "SAV"
    result, elapsed_time, mem_allocated, gc_time, memory_counters = @timed CahnHilliard_SAV(phi0; t_iter=max_it, dt=dt, dt_out=dt_out, m=m, boundary=boundary, printphi=printphi, pathname=pathname)
    open("$(outdir)/Job_specs_updated.csv", "a", lock=false) do f
        writedlm(f, [Dates.format(date_time, "mm/dd/yyyy HH:MM:SS") "Julia" method GridSize epsilon dt "NaN" max_it "NaN" dt_out print boundary pathname elapsed_time (mem_allocated/1e6) SLURM_ID note], ",")
    end
end

if print
    println("Printing out results...")
    t_out = result[1]
    phi_t = result[2]
    mass_t = result[3]
    E_t = result[4]
    open("$(pathname)t_out.csv", "w", lock=false) do f
        writedlm(f, t_out, " ")
    end
    open("$(pathname)mass.csv", "w", lock=false) do f
        writedlm(f, mass_t, " ")
    end
    open("$(pathname)energy.csv", "w", lock=false) do f
        writedlm(f, E_t, " ")
    end
    # t_out = readdlm("$(pathname)t_out.csv", ',', Float64)
    ch_movie_from_file("$(pathname)phi.csv", t_out, GridSize; dtframes=1, filename="$(pathname)movie", filetype="mp4", colorbar_type="default")
end

File: Figure 2/run_spinodal_decomp_HPC.m
function run_spinodal_decomp_HPC(GridSize, boundary, print_results, solver, SLURM_ID, note)
    indir = "../IC";
    outdir = "/project/g_bme-janeslab/SarahG/spinodal_decomp_06_2025"

    n_relax = 4;
    m = 8;
    h = 1/GridSize;
    epsilon = m * h/ (2 * sqrt(2) * atanh(0.9)); 

    dt = 5.5e-7;
    max_it = 20000;

    %run both periodic and neumann with the IC that were smoothed using neumann-BC-smoothing
    init_file = sprintf("%s/initial_phi_%d_smooth_n_relax_%d_%s_neumann.csv",indir,GridSize, n_relax,note);
    phi0 = readmatrix(init_file);
    print_phi = true;
    dt_out = 10;
    ny = GridSize;
    if print_results == "true"
        dt_out = 10
    else
        dt_out = 20000
    end
    pathname = sprintf("%s/out_MATLAB/%s_MATLAB_%d_dt_%.2e_Nx_%d_%s_dt_out_%d%s_",outdir,solver, max_it,dt, GridSize, boundary, dt_out,note);

    % #################################################
    % RUN SAV SOLVER 
    % #################################################
    if solver == "SAV"
        fprintf("Running SAV solver with parameters: %s\n", pathname);
        tol = "NaN";
        solver_iter = "NaN";

        tStart_SAV = tic;
   
        [t_out, phi_t, delta_mass_t, E_t] = CahnHilliard_SAV(phi0,...
                                            t_iter = max_it,...
                                            dt = dt,...
                                            m = m,...
                                            boundary = boundary,...
                                            printphi=print_phi,...
                                            pathname=pathname,...
                                            dt_out = dt_out);
        elapsedTime = toc(tStart_SAV);


    end
    % % #################################################
    % % RUN NMG SOLVER 
    % % #################################################

    if solver == "NMG"
        fprintf("Running NMG solver with parameters: %s\n", pathname);
        tStart_NMG = tic;
        tol = "1e-5"
        solver_iter = "1e4"
        [t_out, phi_t, delta_mass_t, E_t] = CahnHilliard_NMG(phi0,...
                                            t_iter = max_it,...
                                            dt = dt,...
                                            m = m,...
                                            boundary = boundary,...
                                            printphi=print_phi,...
                                            pathname=pathname,...
                                            dt_out = dt_out);
        elapsedTime = toc(tStart_NMG);


    end
    filename = strcat(pathname, "movie");
    fprintf("Creating movie\n");
    if print_results
        ch_movie_from_file(strcat(pathname,"phi.csv"), t_out, ny,filename = filename, filetype = "Motion JPEG AVI")

    writematrix(t_out,sprintf('%st_out.csv', pathname));
    writematrix(delta_mass_t,sprintf('%smass.csv', pathname));
    writematrix(E_t,sprintf('%senergy.csv', pathname));

    fid = fopen(sprintf('%s/Job_specs.csv', outdir), 'a+');
    v = [string(datetime) "MATLAB" solver GridSize epsilon dt tol max_it solver_iter dt_out print_results boundary pathname elapsedTime "NaN" SLURM_ID note];
    fprintf(fid, '%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s\n', v);
    fclose(fid);

end
